Upgrade devtools to 0.1.1 and RSpec to 3.3.1

backus opened the pull request:
- I removed the outdated `Gemfile.devtools` and pointed the Gemfile
  to the newly released devtools gem
 - I updated RSpec to 3.1.1 and updated the specs accordingly.  The
   syntax conversions for RSpec 3 were performed using Transpec. I've
   preserved the conversion summary below.
 - I fixed a few rubocop violations caught by the rubocop upgrade.  I also disabled a few cops that did not  fit the style used in this repo
 - I fixed a few specs to silence rspec warnings which are new to RSpec 3
 - I adjusted the travis build matrix

See the individual commits for more detail

> This conversion is done by Transpec 3.1.1 with the following command:
>    `transpec --keep its,oneliner`
>
> * 37 conversions
>     from: `obj.stub(:message)`
>       to: `allow(obj).to receive(:message)`
> 
> * 20 conversions
>     from: `obj.should_receive(:message)`
>       to: `expect(obj).to receive(:message)`
> 
> * 16 conversions
>     from: `== expected`
>       to: `eq(expected)`
> 
> For more details: https://github.com/yujinakayama/transpec#supported-conversions

backus commented:
Anima's dependency on >= 2.1 breaks the build.  I can add an updated travis matrix

backus commented:
Ok I'm not quite sure why Travis is still failing now. It looks like it is saying there is a conflict between rspec versions it wants to use but I don't know where that conflict would be coming from.

backus commented:
Updating the bundler development dependency fixed the bundling issue. Now new rubocop cops are failing the build since they aren't configured.  I'll try to update the rubocop settings accordingly.

Since I expect to run into a few more build issues I'll split up my opinionated changes (like modifying rubocop settings) in separate commits which I can squash down, modify, or throw out later.

backus commented:
Alright there are still some JRuby build issues which I'll look into (at least for jruby-head).  @dkubb would it be fine to drop everything below 2.1.7 from the travis matrix?

The MRI builds pass except for mutant which finds new mutations now that it is updated.  I'm happy to try and tackle these mutations but it might make sense in the context of a separate PR.

backus commented:
Some builds fail due to a neutral failure quoted below.  I'm going to add this to ignore_subjects for now

> --- Neutral failure ---
> Original code was inserted unmutated. And the test did NOT PASS.
> Your tests do not pass initially or you found a bug in mutant / unparser.
> Subject AST:
> ```
> (def :define
>   (args)
>   (begin
>     (send nil :desc
>       (dstr
>         (str "Measure docs in ")
>         (begin
>           (send
>             (ivar :@config) :path))
>         (str " with yardstick")))
>     (block
>       (send nil :task
>         (ivar :@name))
>       (args)
>       (send nil :yardstick_measure))))
> ```
> Unparsed Source:
> ```ruby
> def define
>   desc("Measure docs in #{@config.path} with yardstick")
>   task(@name) do
>     yardstick_measure
>   end
> end
> ```

mbj commented:
> Some builds fail due to a neutral failure quoted below. I'm going to add this to ignore_subjects for now

JFYI: In the presence of neutrals: You *cannot* know the killed evils are killed for real. So the coverage in presence of neutrals cannot be trusted.

dkubb commented:
> would it be fine to drop everything below 2.1.7 from the travis matrix?

@backus sure. If someone wants it added back in they can create a PR with the fixes and travis additions. I don't use anything except ruby 2.2.3 in production now anyway.

backus commented:
@dkubb This should be ready for review

mbj commented on spec/integration/yardstick/document/file_spec.rb:
```diff
@@ -12,5 +12,5 @@
   it { should be_kind_of(Pathname) }
 
-  it { should == file }
+  it { should eq(file) }
```
I think we should use the most "strongest" matcher to specify the expectation. In that case we should use `should eql(file)`.

`eq` is calling `#==` internally, and `#==` is less "stronger" (contains more semantics) than `#eql?` which is the backend for the `eql` expectation.

mbj commented on spec/integration/yardstick/document/path_spec.rb:
```diff
@@ -22,7 +22,7 @@ def baz(value)
     end
 
-    it { should == 'Foo::Bar#baz' }
+    it { should eq('Foo::Bar#baz') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/integration/yardstick/document/path_spec.rb:
```diff
@@ -40,6 +40,6 @@ def self.baz(value)
     end
 
-    it { should == 'Foo::Bar.baz' }
+    it { should eq('Foo::Bar.baz') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/class_methods/measure_spec.rb:
```diff
@@ -8,10 +8,10 @@
     it 'delegates to Processor' do
       processor = double('processor')
-      Yardstick::Processor.stub(:new).with(instance_of(Yardstick::Config)) do
+      allow(Yardstick::Processor).to receive(:new).with(instance_of(Yardstick::Config)) do
```
Instead of using the block to define the return value, I think that `and_return(processor)` would use a less powerful primitive.

mbj commented on spec/unit/yardstick/class_methods/measure_spec.rb:
```diff
@@ -23,8 +23,8 @@
       processor = double('processor')
-      Yardstick::Processor.stub(:new).with(config) { processor }
-      processor.should_receive(:process)
+      allow(Yardstick::Processor).to receive(:new).with(config) { processor }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/class_methods/measure_string_spec.rb:
```diff
@@ -10,10 +10,10 @@
     it 'delegates to Processor' do
       processor = double('processor')
-      Yardstick::Processor.stub(:new).with(instance_of(Yardstick::Config)) do
+      allow(Yardstick::Processor).to receive(:new).with(instance_of(Yardstick::Config)) do
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/class_methods/measure_string_spec.rb:
```diff
@@ -25,8 +25,8 @@
       processor = double('processor')
-      Yardstick::Processor.stub(:new).with(config) { processor }
-      processor.should_receive(:process_string).with(string)
+      allow(Yardstick::Processor).to receive(:new).with(config) { processor }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/config/output_spec.rb:
```diff
@@ -10,6 +10,6 @@
     it { should be_instance_of(Yardstick::ReportOutput) }
 
-    its(:to_s) { should == 'measurements/report.txt' }
+    its(:to_s) { should eq('measurements/report.txt') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989997

mbj commented on spec/unit/yardstick/config/output_writer_spec.rb:
```diff
@@ -15,6 +15,6 @@
     it { should be_a(Yardstick::ReportOutput) }
 
-    its(:to_s) { should == 'tmp/*.rb' }
+    its(:to_s) { should eq('tmp/*.rb') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989997

mbj commented on spec/unit/yardstick/config/path_spec.rb:
```diff
@@ -8,6 +8,6 @@
     let(:options) { {} }
 
-    it { should == 'lib/**/*.rb' }
+    it { should eq('lib/**/*.rb') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989997

mbj commented on spec/unit/yardstick/config/path_writer_spec.rb:
```diff
@@ -7,5 +7,5 @@
   before { subject.path = '/tmp/*.rb' }
 
-  its(:path) { should == '/tmp/*.rb' }
+  its(:path) { should eq('/tmp/*.rb') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989997

mbj commented on spec/unit/yardstick/document/api_predicate_spec.rb:
```diff
@@ -8,7 +8,7 @@
 
   before do
-    docstring.stub(:tag).with('api') { double(:text => 'private') }
+    allow(docstring).to receive(:tag).with('api') { double(:text => 'private') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/document/class_methods/measure_spec.rb:
```diff
@@ -14,7 +14,7 @@
   before do
     registered_rules.each do |rule_class|
-      rule_class.should_receive(:coerce).with(document, config)
+      expect(rule_class).to receive(:coerce).with(document, config)
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

backus replied:
This comment didn't hide since the change is on the following line

mbj commented on spec/unit/yardstick/document/tag_text_spec.rb:
```diff
@@ -10,7 +10,7 @@
 
   before do
-    docstring.stub(:tag).with(name) { yard_tag }
+    allow(docstring).to receive(:tag).with(name) { yard_tag }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/document/tag_types_spec.rb:
```diff
@@ -11,7 +11,7 @@
 
   before do
-    docstring.stub(:tag).with(name) { yard_tag }
+    allow(docstring).to receive(:tag).with(name) { yard_tag }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/measurement/initialize_spec.rb:
```diff
@@ -11,7 +11,7 @@
 
     before do
-      Yardstick::Config.stub(:coerce).with(options) { config }
+      allow(Yardstick::Config).to receive(:coerce).with(options) { config }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/measurement/yardstick_measure_spec.rb:
```diff
@@ -13,10 +13,10 @@
-    Yardstick.stub(:measure).with(config) { measurements }
-    report_writer.stub(:write).and_yield(io)
-    measurements.should_receive(:puts).with(io)
+    allow(Yardstick::Config).to receive(:coerce).with(options) { config }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/measurement/yardstick_measure_spec.rb:
```diff
@@ -13,10 +13,10 @@
-    report_writer.stub(:write).and_yield(io)
-    measurements.should_receive(:puts).with(io)
+    allow(Yardstick::Config).to receive(:coerce).with(options) { config }
+    allow(Yardstick).to receive(:measure).with(config) { measurements }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/verify/initialize_spec.rb:
```diff
@@ -11,7 +11,7 @@
 
     before do
-      Yardstick::Config.stub(:coerce).with(options) { config }
+      allow(Yardstick::Config).to receive(:coerce).with(options) { config }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/verify/verify_measurements_spec.rb:
```diff
@@ -12,8 +12,8 @@
   before do
-    Yardstick::Config.stub(:coerce).with(options) { config }
-    Yardstick.stub(:measure).with(config) { measurements }
+    allow(Yardstick::Config).to receive(:coerce).with(options) { config }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/verify/verify_measurements_spec.rb:
```diff
@@ -12,8 +12,8 @@
-    Yardstick::Config.stub(:coerce).with(options) { config }
-    Yardstick.stub(:measure).with(config) { measurements }
+    allow(Yardstick::Config).to receive(:coerce).with(options) { config }
+    allow(Yardstick).to receive(:measure).with(config) { measurements }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rule/class_methods/coerce_spec.rb:
```diff
@@ -10,7 +10,7 @@
 
   before do
-    config.stub(:for_rule).with(described_class) { rule_config }
+    allow(config).to receive(:for_rule).with(described_class) { rule_config }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/measurement/description_spec.rb:
```diff
@@ -8,5 +8,5 @@
   let(:document) { DocumentMock.new        }
 
-  it { should == ValidRule.description }
+  it { should eq(ValidRule.description) }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -15,19 +15,19 @@
       let(:rule) { ValidRule.new(document) }
 
-      it { should == '' }
+      it { should eq('') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -15,19 +15,19 @@
       let(:rule) { NotValidatableRule.new(document) }
 
-      it { should == '' }
+      it { should eq('') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -15,19 +15,19 @@
       let(:rule) { InvalidRule.new(document) }
 
-      it { should == "(stdin):2: Foo#bar: not successful\n" }
+      it { should eq("(stdin):2: Foo#bar: not successful\n") }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -44,19 +44,19 @@
       let(:rule) { ValidRule.new(document) }
 
-      it { should == '' }
+      it { should eq('') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -44,19 +44,19 @@
       let(:rule) { NotValidatableRule.new(document) }
 
-      it { should == '' }
+      it { should eq('') }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/measurement/puts_spec.rb:
```diff
@@ -44,19 +44,19 @@
       let(:rule) { InvalidRule.new(document) }
 
-      it { should == "(stdin):2: Foo#bar: not successful\n" }
+      it { should eq("(stdin):2: Foo#bar: not successful\n") }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/processor/process_string_spec.rb:
```diff
@@ -10,8 +10,8 @@
   it 'measures specified string' do
-    Yardstick::Parser.should_receive(:parse_string).with(string) { documents }
-    documents.should_receive(:measure).with(config)
+    expect(Yardstick::Parser).to receive(:parse_string).with(string) { documents }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/processor/process_spec.rb:
```diff
@@ -10,8 +10,8 @@
   it 'measures files specified in the config' do
-    Yardstick::Parser.should_receive(:parse_paths).with(['foo/bar.rb']) { documents }
-    documents.should_receive(:measure).with(config)
+    expect(Yardstick::Parser).to receive(:parse_paths).with(['foo/bar.rb']) { documents }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39990084

mbj commented on spec/unit/yardstick/rake/verify/verify_measurements_spec.rb:
```diff
@@ -42,7 +42,7 @@
     it 'outputs coverage' do
-      expect { measure }.to raise_error
+      expect { measure }.to raise_error(RuntimeError)
       expect(@output).to eq("YARD-Coverage: 43.4% (threshold: 90%)\n")
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/rake/verify/verify_measurements_spec.rb:
```diff
@@ -56,7 +56,7 @@
     it 'outputs coverage' do
-      expect { measure }.to raise_error
+      expect { measure }.to raise_error(RuntimeError)
       expect(@output).to eq("YARD-Coverage: 99.8% (threshold: 90%)\n")
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented on spec/unit/yardstick/report_output/to_s_spec.rb:
```diff
@@ -8,5 +8,5 @@
   let(:path)     { '/foo/bar'     }
 
-  it { should == path }
+  it { should eq(path) }
```
Instance of https://github.com/dkubb/yardstick/pull/27/files#r39989956

mbj commented:
@backus I did a code only pass over the diff and noticed 2 class of problems I'd like to see addressed. On the first instance I left a verbose description, on each instance a link to the description. The idea is that if you fix the instances, the "instance comment" will not be visible in the PR view anymore, as a confirmation you catched all instance of a certain class. I realize you did the transform with transproc, but still I think we should always go for the best possible expectations.

OT: Conceptually I think github should provide better review tooling.

Also I notice that this code base still uses 1.8 hashes, I think we should add a dedicated commit to use 1.9 style hashes where possible.

backus commented:
Fixed the mentioned problems @mbj and also used rubocop to update the hash syntax.  The instance comments were actually really helpful for knowing if I addressed everything.  They were also helpful for getting to the issues quickly (using [jq](https://stedolan.github.io/jq/manual/)):

```sh
curl 'https://api.github.com/repos/dkubb/yardstick/pulls/27/comments' | jq 'map(.path) | .[]' -r | sort | uniq | xargs -L1 open
```

dkubb commented on spec/unit/yardstick/class_methods/measure_spec.rb:
```diff
@@ -8,10 +8,11 @@
+
+      allow(Yardstick::Processor)
+        .to receive(:new).with(config).and_return(processor)
+      expect(processor).to receive(:process)
```
Below in spec/unit/yardstick/class_methods/measure_string_spec.rb you have basically the same expression, yet totally different formatting. Assuming the line length isn't too long, I prefer the other formatting.

backus replied:
The rubocop config says this repo wants to keep lines under 80 chars. Is this still the preference?

dkubb replied:
Yes, if the line length is too much, then I think it's fine to leave it as-is.

dkubb commented on spec/unit/yardstick/config/set_defaults_spec.rb:
```diff
@@ -24,10 +24,10 @@
-        :verbose => false,
-        :path => 'tmp/*.rb',
-        :require_exact_threshold => false
+        threshold: 15,
```
I would tend to vertically align hash values.

dkubb commented on spec/unit/yardstick/document/file_spec.rb:
```diff
@@ -5,10 +5,10 @@
   it { should be_kind_of(Pathname) }
 
-  its(:to_s) { should == '/foo/bar.rb' }
+  its(:to_s) { should eq('/foo/bar.rb') }
```
Does `#eql(...)` work here?

backus replied:
Yup

dkubb commented on spec/unit/yardstick/document/path_spec.rb:
```diff
@@ -5,8 +5,8 @@
+  let(:docstring) { double('docstring', object: object) }
+  let(:object)    { double('object', path: 'Foo#bar')   }
 
   it { should eq('Foo#bar') }
```
Not that it is required for this PR, given that this exists in the source and there was one more `eq` -> `eql` found, I wonder if we should just do a find/replace of all `'eq('` strings into `'eql('`?

backus replied:
I replaced the remaining `eq(` locally actually.  Tests still pass

dkubb commented on .travis.yml:
```diff
@@ -5,19 +5,12 @@ env:
-  - jruby-19mode
-  - jruby-head
-  - rbx-19mode
+  - '2.1.6'
```
Should we test `2.1.7` instead since it's the latest in the `2.1.x` series?

dkubb replied:
My mistake. In https://github.com/backus/yardstick/commit/1dc4c6c928089fd80d166172d9e35a7f95af7f05 it's explicitly stated that 2.1.6 was used due to an error in Parser.

dkubb commented:
@backus I finished a pass over this code. I think once the flagged issues are resolved I'm ready to merge.

dkubb commented:
@backus looks great. I will merge once everything is green.

dkubb commented:
@backus thanks for your contribution!

backus commented:
:)
